### PR TITLE
Update crypo to latest(0.17.0) for solve Terrapin Attack vulnerabilities(CVE-2023-48795, CVE-2023-46445, CVE-2023-46446)

### PR DESCRIPTION
**Pull request by heat1024**
## Overview

To resolve Terrapin Attack vulnerabilities, This PR makes crypo to latest.
Want to use crypto 0.17.0

- [CVE-2023-48795](https://github.com/advisories/GHSA-45x7-px36-x8w8)
- [CVE-2023-46445](https://github.com/advisories/GHSA-cfc2-wr2v-gxm5)
- [CVE-2023-46446](https://github.com/advisories/GHSA-c35q-ffpf-5qpm)

There is no conflicts with previous codes in this repository.

## Diff

Append set PublicKeyAuthAlgorithms variable in ServerConfig. It is a new variable for set auth algorithm by config.
Auth algorithms variable was set in common.go before, but now, It is moved to ServerConfig.

```Golang
	if len(fullConf.PublicKeyAuthAlgorithms) == 0 {
		fullConf.PublicKeyAuthAlgorithms = supportedPubKeyAuthAlgos
	} else {
		for _, algo := range fullConf.PublicKeyAuthAlgorithms {
			if !contains(supportedPubKeyAuthAlgos, algo) {
				c.Close()
				return nil, fmt.Errorf("ssh: unsupported public key authentication algorithm %s", algo)
			}
		}
	}
```

---

The other changes are just merged golang/crypto's master.

### CODE REVIEW

File: ssh/common.go [hunk 1]
```diff
@@ -131,8 +147,6 @@ var supportedPubKeyAuthAlgos = []string{
 	KeyAlgoDSA,
 }
 
-var supportedPubKeyAuthAlgosList = strings.Join(supportedPubKeyAuthAlgos, ",")
```
Comment (heat1024) [inline ssh/common.go:134]: It is removed supportedPubKeyAuthAlgos for handshake with client.

File: ssh/handshake.go [hunk 2]
```diff
@@ -642,16 +695,21 @@ func (t *handshakeTransport) enterKeyExchange(otherInitPacket []byte) error {
 
 	// On the server side, after the first SSH_MSG_NEWKEYS, send a SSH_MSG_EXT_INFO
 	// message with the server-sig-algs extension if the client supports it. See
-	// RFC 8308, Sections 2.4 and 3.1.
+	// RFC 8308, Sections 2.4 and 3.1, and [PROTOCOL], Section 1.9.
 	if !isClient && firstKeyExchange && contains(clientInit.KexAlgos, "ext-info-c") {
+		supportedPubKeyAuthAlgosList := strings.Join(t.publicKeyAuthAlgorithms, ",")
```
Comment (heat1024) [inline ssh/handshake.go:700]: handshake use `publicKeyAuthAlgorithms` from `ServerConfig.PublicKeyAuthAlgorithms` now.
So, we must set it after set server config in proxy.go

File: ssh/proxy.go [hunk 3]
```diff
@@ -452,6 +452,17 @@ func NewDownstreamConn(c net.Conn, config *ServerConfig) (*connection, error) {
 	fullConf := *config
 	fullConf.SetDefaults()
 
+	if len(fullConf.PublicKeyAuthAlgorithms) == 0 {
+		fullConf.PublicKeyAuthAlgorithms = supportedPubKeyAuthAlgos
```
Comment (heat1024) [inline ssh/proxy.go:456]: Set PublicKeyAuthAlgorithms by default(in common.go) after set default server config.

File: ssh/common.go [hunk 4]
```diff
@@ -131,8 +147,6 @@ var supportedPubKeyAuthAlgos = []string{
 	KeyAlgoDSA,
```
Comment (heat1024) [inline ssh/common.go:147]: These default supportedPubKeyAuthAlgorithms are same with privious version.

**Comment by heat1024** [comment]
@tsurubee
May I get a review for these PR?

After merge it, Please update https://github.com/tsurubee/sshr 's go modules also. 🙏

**Comment by tsurubee** [comment]
@heat1024 
LGTM! Thank you for your PR!
Could you also send me sshr's PR if possible?

**Comment by heat1024** [comment]
OK I will 😄

**Comment by tsurubee** [comment]
Thank you!
I will check later!

**Pull request merged**